ddtrace/internal/datadog/profiling/dd_wrapper/src/code_provenance.cpp

```diff
@@ -0,0 +1,178 @@
+#include "code_provenance.hpp"
+
+#include <memory>
+#include <mutex>
+#include <set>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <vector>
+
+namespace Datadog {
+
+void
+Datadog::CodeProvenance::postfork_child()
+{
+    get_instance().mtx.~mutex();            // Destroy the mutex
+    new (&get_instance().mtx) std::mutex(); // Recreate the mutex
+    get_instance().reset();                 // Reset the state
+}
+
+void
+Datadog::CodeProvenance::set_enabled(bool enable)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    enabled = enable;
+}
+
+bool
+Datadog::CodeProvenance::is_enabled()
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    return enabled;
+}
+
+void
+Datadog::CodeProvenance::set_runtime_version(std::string_view _runtime_version)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    this->runtime_version = _runtime_version;
+}
+
+void
+Datadog::CodeProvenance::set_stdlib_path(std::string_view _stdlib_path)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    this->stdlib_path = _stdlib_path;
+}
+
+void
+CodeProvenance::add_packages(std::unordered_map<std::string_view, std::string_view> distributions)
+{
+    std::lock_guard<std::mutex> lock(mtx);
```

Comment:
If `enabled` was atomic, it'd be possible to move the lock guard to after the check.

Reply:
Updated to use `atomic`